src/govuk/i18n.mjs

```diff
@@ -15,19 +15,56 @@ function I18n (translations) {
  * returns the appropriate string.
  *
  * @param    {String}  lookupKey  - The lookup key of the string to use.
+ * @param    {Object}  options    - Any options passed with the translation string, e.g: for string interpolation.
  * @returns  {String}             - The appropriate translation string.
  */
-I18n.prototype.t = function (lookupKey) {
+I18n.prototype.t = function (lookupKey, options) {
   if (!lookupKey) {
     // Print a console error if no lookup key has been provided
     throw new Error('i18n lookup key missing.')
   } else if (lookupKey in this.translations) {
-    // Return the translation string if it exists in the object
-    return this.translations[lookupKey]
+    // Fetch the translation string for that lookup key
+    var translationString = this.translations[lookupKey]
+
+    // Check for ${} placeholders in the translation string
+    if (translationString.match(/\${(.\S+)}/)) {
+      if (!options) {
+        throw new Error('i18n: cannot replace placeholders in string if no option data provided')
+      }
+
+      return this.replacePlaceholders(translationString, options)
+    } else {
+      return translationString
+    }
   } else {
     // Otherwise, return the lookup key itself as the fallback
     return lookupKey
   }
 }
 
+/**
+ * Takes a translation string with placeholders, and replaces the placeholders
+ * with the provided data
+ *
+ * @param    {String}  translationString  - The translation string
+ * @param    {Object}  options    - Any options passed with the translation string, e.g: for string interpolation.
+ * @returns  {String}             - The translation string to output, with ${} placeholders replaced
+ */
+I18n.prototype.replacePlaceholders = function (translationString, options) {
+  var placeholderRegex = RegExp(/\${(.\S+)}/, 'g')
+  var placeholderMatch
+
+  // Use `exec` for fetching regex matches, as matchAll() is not supported in IE
+  while ((placeholderMatch = placeholderRegex.exec(translationString)) !== null) {
+    var placeholderKey = placeholderMatch[1]
+    if (!options[placeholderKey]) {
```

Comment:
I think this will catch falsy values, especially `''`, as well, is that:

1. something we want to happen, in which case we might need to reword the error message as I'm not sure people will think of checking that the value they have in their hash is not falsy when reading "no data found"
2. something we don't want to happen (the only two cases I could imagine is passing an empty string or `false`), in which case we'll need to change the check to `placeholderKey in options`

I'd edge towards 2 as enabling empty values allows to interpolate things like "Some content ${optionalAdditionalContent}", but I'm obviously young to the issue so this may have been discusses ahead of time.

Reply:
Good catch. I hinted at this in my comment about tests, but probably should have called it out explicitly 😄 

Agree we should support replacing with empty strings, and other falsey things like `0` 👍🏻 

> in which case we'll need to change the check to placeholderKey in options

I think this will match names of properties in the prototype chain (e.g. `'hasOwnProperty' in {}` is `true`), but using `hasOwnProperty` should work I think?

Reply:
Thanks @romaricpascal @36degrees , really good catch!